routers/api/v1/repo/pull.go

```diff
@@ -743,29 +742,33 @@ func MergePullRequest(ctx *context.APIContext) {
 		}
 	}
 
-	if pr.Issue.IsClosed {
-		ctx.NotFound()
-		return
-	}
-
-	allowedMerge, err := pull_service.IsUserAllowedToMerge(pr, ctx.Repo.Permission, ctx.Doer)
-	if err != nil {
-		ctx.Error(http.StatusInternalServerError, "IsUSerAllowedToMerge", err)
-		return
-	}
-	if !allowedMerge {
-		ctx.Error(http.StatusMethodNotAllowed, "Merge", "User not allowed to merge PR")
-		return
-	}
+	manuallMerge := repo_model.MergeStyle(form.Do) == repo_model.MergeStyleManuallyMerged
+	force := form.ForceMerge != nil && *form.ForceMerge
 
-	if pr.HasMerged {
-		ctx.Error(http.StatusMethodNotAllowed, "PR already merged", "")
+	if err := pull_service.CheckPullProtection(ctx, ctx.Doer, &ctx.Repo.Permission, pr, manuallMerge, force); err != nil {
+		if pull_service.IsErrIsClosed(err) {
+			ctx.NotFound()
+		} else if pull_service.IsErrUserNotAllowedToMerge(err) {
+			ctx.Error(http.StatusMethodNotAllowed, "Merge", "User not allowed to merge PR")
+		} else if pull_service.IsErrHasMerged(err) {
+			ctx.Error(http.StatusMethodNotAllowed, "PR already merged", "")
+		} else if pull_service.IsErrIsWorkInProgress(err) {
+			ctx.Error(http.StatusMethodNotAllowed, "PR is a work in progress", "Work in progress PRs cannot be merged")
+		} else if pull_service.IsErrNotMergableState(err) {
+			ctx.Error(http.StatusMethodNotAllowed, "PR not in mergeable state", "Please try again later")
+		} else if models.IsErrNotAllowedToMerge(err) {
+			ctx.Error(http.StatusMethodNotAllowed, "PR is not ready to be merged", err)
+		} else if asymkey_service.IsErrWontSign(err) {
+			ctx.Error(http.StatusMethodNotAllowed, fmt.Sprintf("Protected branch %s requires signed commits but this merge would not be signed", pr.BaseBranch), err)
```

Comment:
It seems that many calls to `ctx.Error` should be fine tuned.

`func (ctx *APIContext) Error(status int, title string, obj interface{})`

The `ctx.Error` is not well designed. the `title` is for log only, it's never shown to end users.

And exposing `err(error)` to end users is dangerous, it many contain sensitive information (for example, the git.Command.RunXxx may return err with stderr, which may contain server internal information)

Reply:
this was just the original code - I didn't touch that as it was not the porpoise of this pull to change behavior